File: bidding-engine/src/main/java/com/omnibus/bidding/rules/RuleCatalogService.java
package com.omnibus.bidding.rules;

import com.omnibus.bidding.application.ManagedRuleStore;
import com.omnibus.bidding.domain.ManagedRuleDefinition;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import org.springframework.core.io.Resource;
import org.springframework.core.io.support.PathMatchingResourcePatternResolver;
import org.springframework.stereotype.Service;

@Service
public class RuleCatalogService {

    private final ManagedRuleStore managedRuleStore;
    private final PathMatchingResourcePatternResolver resourcePatternResolver = new PathMatchingResourcePatternResolver();

    public RuleCatalogService(ManagedRuleStore managedRuleStore) {
        this.managedRuleStore = managedRuleStore;
    }

    public List<ManagedRuleDefinition> listAllRules() {
        List<ManagedRuleDefinition> ruleDefinitions = new ArrayList<>();
        ruleDefinitions.addAll(loadBundledRules());
        ruleDefinitions.addAll(managedRuleStore.listManagedRules());
        return ruleDefinitions;
    }

    public List<ManagedRuleDefinition> listManagedRules() {
        return managedRuleStore.listManagedRules();
    }

    private List<ManagedRuleDefinition> loadBundledRules() {
        try {
            Resource[] resources = resourcePatternResolver.getResources("classpath*:rules/*.drl");
            List<ManagedRuleDefinition> bundledRules = new ArrayList<>();

            for (Resource resource : resources) {
                bundledRules.add(new ManagedRuleDefinition(
                    resource.getFilename(),
                    "src/main/resources/rules/" + resource.getFilename(),
                    false,
                    new String(resource.getInputStream().readAllBytes(), StandardCharsets.UTF_8)
                ));
            }

            return bundledRules;
        } catch (IOException exception) {
            throw new IllegalStateException("Unable to read bundled rules", exception);
        }
    }
}

File: services/bidding-engine/src/main/kotlin/pl/writeonly/omnibus/bidding/rules/RuleCatalogService.kt
package pl.writeonly.omnibus.bidding.rules

import com.omnibus.bidding.application.ManagedRuleStore
import com.omnibus.bidding.domain.ManagedRuleDefinition
import org.springframework.core.io.support.PathMatchingResourcePatternResolver
import org.springframework.stereotype.Service
import java.io.IOException
import java.nio.charset.StandardCharsets

@Service
class RuleCatalogService(
    private val managedRuleStore: ManagedRuleStore,
) {
    private val resourcePatternResolver = PathMatchingResourcePatternResolver()

    fun listAllRules(): List<ManagedRuleDefinition> =
        buildList {
            addAll(loadBundledRules())
            addAll(managedRuleStore.listManagedRules())
        }

    fun listManagedRules(): List<ManagedRuleDefinition> = managedRuleStore.listManagedRules()

    private fun loadBundledRules(): List<ManagedRuleDefinition> =
        try {
            val resources = resourcePatternResolver.getResources("classpath*:rules/*.drl")
            resources.map { resource ->
                val fileName = resource.filename
                    ?: throw IllegalStateException("Unable to read bundled rules (missing filename)")
                ManagedRuleDefinition(
                    fileName,
                    "src/main/resources/rules/$fileName",
                    false,
                    String(resource.inputStream.readAllBytes(), StandardCharsets.UTF_8),
                )
            }
        } catch (exception: IOException) {
            throw IllegalStateException("Unable to read bundled rules", exception)
        }
}

